raft/server.cc

```diff
@@ -1205,49 +1232,70 @@ future<read_barrier_reply> server_impl::get_read_idx(server_id leader, seastar::
     }
 }
 
-future<> server_impl::read_barrier(seastar::abort_source* as) {
-    server_id leader = _fsm->current_leader();
+template <typename AsyncAction>
+SEASTAR_CONCEPT( requires requires (server_id& leader, AsyncAction aa) {
+    { aa(leader) } -> std::same_as<future<stop_iteration>>;
+} )
+future<> server_impl::do_on_leader_with_retries(seastar::abort_source* as, AsyncAction action) {
+    server_id leader = _fsm->current_leader(), prev_leader{};
 
-    logger.trace("[{}] read_barrier start", _id);
-    index_t read_idx;
-
-    while (read_idx == index_t{}) {
+    do {
         if (as && as->abort_requested()) {
             throw request_aborted();
         }
         check_not_aborted();
-        logger.trace("[{}] read_barrier forward to  {}", _id, leader);
         if (leader == server_id{}) {
             co_await wait_for_leader(as);
             leader = _fsm->current_leader();
-        } else {
-            auto applied = _applied_idx;
-            read_barrier_reply res;
-            bool need_retry = false;
-            try {
-                res = co_await get_read_idx(leader, as);
-            } catch (const transport_error& e) {
-                logger.trace("[{}] read_barrier on {} resulted in {}; retrying", _id, leader, e);
-                need_retry = true;
-            }
-            if (need_retry) {
-                leader = server_id{};
-                co_await yield();
-                continue;
-            }
-            if (std::holds_alternative<std::monostate>(res)) {
-                // the leader is not ready to answer because it did not
-                // committed any entries yet, so wait for any entry to be
-                // committed (if non were since start of the attempt) and retry.
-                logger.trace("[{}] read_barrier leader not ready", _id);
-                co_await wait_for_apply(++applied, as);
-            } else if (std::holds_alternative<raft::not_a_leader>(res)) {
-                leader = std::get<not_a_leader>(res).leader;
-            } else {
-                read_idx = std::get<index_t>(res);
-            }
+            continue;
         }
-    }
+        if (prev_leader && leader == prev_leader) {
+            // This is to protect against tight loop in case we didn't get
+            // any new information about the current leader.
+            // This can happen if the server responds with a transient_error with
+            // an empty leader and the current node has not yet learned the new leader.
+            // We neglect an excessive delay if the newly elected leader is the same as
+            // the previous one, this supposed to be a rare.
+            co_await wait_for_next_tick(as);
+            prev_leader = leader = server_id{};
+            continue;
```

Comment:
Could be easier to read without the `continue`:
here assign `leader = server_id{}`, and the next statement will assign `prev_leader = leader`.

But I have a suspicion that this code is written with the assumption that `continue` will not evaluate `co_await action(leader)` (but it will).